Bluetooth: controller: Extended/Periodic Advertising time reservations

cvinayak opened the pull request:
Add Extended and Periodic Advertising radio event time
reservations.

~TODO: Periodic Advertising time reservation update is work in progress, and can be added to this PR or separate PR.~ Done

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Extended/Periodic Advertising time reservations

Add Extended and Periodic Advertising radio event time
reservations.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: nRF5: Use conservative Tx-Rx turnaround timing

Use a conservative turnaround time of 150us covering slower
nRF51 and faster nRF52 CPU that influence the turnaround
timing.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1232,7 +1237,11 @@ uint8_t ll_adv_enable(uint8_t enable)
+					MAX(EVENT_MAFS_US,
+					    EVENT_OVERHEAD_START_US) -
+					EVENT_OVERHEAD_START_US +
+					(EVENT_TICKER_RES_MARGIN_US << 1));
```
Please put a comment why do you add `(EVENT_TICKER_RES_MARGIN_US << 1));` to the anchor point for auxiliary PDU.

cvinayak replied:
done.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1245,9 +1254,12 @@ uint8_t ll_adv_enable(uint8_t enable)
+						MAX(EVENT_MAFS_US,
+						    EVENT_OVERHEAD_START_US) -
+						EVENT_OVERHEAD_START_US +
+						(EVENT_TICKER_RES_MARGIN_US << 1));
```
Please put a comment why do you add `(EVENT_TICKER_RES_MARGIN_US << 1));` to the anchor point for sync PDU.

cvinayak replied:
done.

cvinayak committed:
Bluetooth: Controller: Refactor out Periodic Advertising time calc

Refactor out Periodic Advertising time calculations so that
it can be reused when PDU length is updated as part of
advertising data updates.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Refactor out advertising time calculations

Refactor out advertising time calculations so that it can be
reused when PDU length is updated as part of advertising
data and scan response data updates.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Refactor out Extended Advertising time calc

Refactor out Extended Advertising time calculations so that
it can be reused when PDU length is updated as part of
advertising data and scan response data updates.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Legacy advertising time reservation

Implementation to update legacy advertising time
reservations when advertising and scan response data are
updated.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Extended advertising time reservation

Implementation to update Extended Advertising time
reservations when advertising and scan response data are
updated.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Periodic advertising time reservation

Implementation to update Periodic Advertising time
reservations when advertising data and/or CTE length is
updated.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Update DID when scan response data removed

Update DID value when scan response data is remove, length
set to 0.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Export primary adv channel event time calc

Export the function calculating the primary channel
advertising event time reservation, so that Extended
Advertising implementation can reuse it.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Disallow Periodic Adv Enable if already enabled

Add check for already enabled and started Periodic
Advertising when HCI LE Set Periodic Advertising Parameters
command is called.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix Extended and Periodic Coded PHY time

Fix Extended and Periodic Coded PHY time reservation. The
define earlier used does not have calculations for Coded
PHY.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_vendor.h:
```diff
@@ -27,5 +27,7 @@
 
-/* TODO - fix up numbers re. HW */
-#define EVENT_RX_TX_TURNAROUND(phy)  ((phy) == 1?100:((phy) == 2 ? 80:900))
+/* Using conservative turnaround time, considering slowest CPU timings that is
+ * required to meet tIFS of 150 us.
+ */
+#define EVENT_RX_TX_TURNAROUND(phy)   150
```
How come this isn't related to the `phy` value anymore? The old maximum value was seemingly 900, so relative to that, this isn't very conservative.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1118,7 +1064,7 @@ uint8_t ll_adv_enable(uint8_t enable)
 		_radio.advertiser.scan_window_ms = scan_window;
 
-		interval_min_us = slot_us + (scan_delay + scan_window) * 1000;
+		interval_min_us = time_us + (scan_delay + scan_window) * 1000;
```
```
interval_min_us = time_us + (scan_delay + scan_window) * USEC_PER_MSEC
```

Thalley reviewed (commented):
Two minor comments

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -245,9 +245,7 @@ uint8_t ll_adv_aux_sr_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref, ui
-		lll_adv_scan_rsp_enqueue(&adv->lll, idx);
-
-		return 0;
+		goto sr_data_set_did_update;
```
What is DID?

cvinayak replied:
Advertising Data ID (DID).

Thalley reviewed (approved):
LGTM, but I probably don't have the full insight into saying whether all the changes are correct or not.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -1025,10 +1067,76 @@ static inline void aux_release(struct ll_adv_aux_set *aux)
+				   TARGETA_SIZE), lll->phy_s);
+
+		time_us += EVENT_IFS_MAX_US * 2 + conn_req_us + conn_rsp_us;
+	} else if (pdu->adv_ext_ind.adv_mode & BT_HCI_LE_ADV_PROP_SCAN) {
```
This may lead to error in if reserved value (0b11) is changed to non-reserved in BT Core spec. releases. More defense approach would be to use regular comparison instead of bitwise-and.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1754,6 +1810,61 @@ static inline struct ll_adv_set *is_disabled_get(uint8_t handle)
+
+			time_us += (BYTES2US(adv_size, PHY_1M) +
+				    EVENT_IFS_MAX_US + rx_to_us +
+				    rxtx_turn_us) * (adv_chn_cnt-1) +
```
```suggestion
				    rxtx_turn_us) * (adv_chn_cnt - 1) +
```

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1718,6 +1729,51 @@ const uint8_t *ull_adv_pdu_update_addrs(struct ll_adv_set *adv,
+		ticks_minus = 0U;
+		ticks_plus = time_ticks - adv->ull.ticks_slot;
+	} else {
+		return 0;
```
```suggestion
		return BT_HCI_ERR_SUCCESS;
```

To stay consistent with line 1769.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1718,6 +1729,51 @@ const uint8_t *ull_adv_pdu_update_addrs(struct ll_adv_set *adv,
+
+	adv->ull.ticks_slot = time_ticks;
+
+	return 0;
```
```suggestion
	return BT_HCI_ERR_SUCCESS;
```

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -1025,10 +1067,76 @@ static inline void aux_release(struct ll_adv_aux_set *aux)
+		ticks_minus = 0U;
+		ticks_plus = time_ticks - aux->ull.ticks_slot;
+	} else {
+		return 0;
```
```suggestion
		return BT_HCI_ERR_SUCCESS;
```

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -1025,10 +1067,76 @@ static inline void aux_release(struct ll_adv_aux_set *aux)
+
+	aux->ull.ticks_slot = time_ticks;
+
+	return 0;
```
```suggestion
	return BT_HCI_ERR_SUCCESS;
```

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -639,40 +657,59 @@ uint16_t ull_adv_sync_lll_handle_get(struct lll_adv_sync *lll)
+
+	ARG_UNUSED(tmp_idx);
+
+	return 0;
```
```suggestion
	return BT_HCI_ERR_SUCCESS;
```

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -832,6 +844,68 @@ uint8_t ull_adv_sync_pdu_alloc(struct ll_adv_set *adv,
+		ticks_minus = 0U;
+		ticks_plus = time_ticks - sync->ull.ticks_slot;
+	} else {
+		return 0;
```
```suggestion
		return BT_HCI_ERR_SUCCESS;
```

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -832,6 +844,68 @@ uint8_t ull_adv_sync_pdu_alloc(struct ll_adv_set *adv,
+
+	sync->ull.ticks_slot = time_ticks;
+
+	return 0;
```
```suggestion
	return BT_HCI_ERR_SUCCESS;
```

carlescufi commented:
@cvinayak please address review comments

cvinayak committed:
Bluetooth: Controller: Use BT_HCI_ERR_SUCCESS return and other changes

Apply suggestions from code review, and change to using
BT_HCI_ERR_SUCCESS instead of returning 0.

Co-authored-by: Piotr Pryga <piotr.pryga@nordicsemi.no>
Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

ppryga-nordic reviewed (approved):
Generally LGTM.

There is one thing that worries me a little bit and it should he handled in future:
https://github.com/zephyrproject-rtos/zephyr/blob/5b7f52d7ad46b8edd3752e00bd9a1178a5e0d402/subsys/bluetooth/controller/ll_sw/ull_adv_sync.c#L672-L674

We have to find some more defensive approach for such situations. 
This is hidden restriction imposed on the function. It can be find only if one looks into function implementation. It would be better to avoid misuses of functions like this one gracefully.

cvinayak commented:
> We have to find some more defensive approach for such situations.
> This is hidden restriction imposed on the function. It can be find only if one looks into function implementation. It would be better to avoid misuses of functions like this one gracefully.

Good point, I will refactor this function in a follow up PR and let the caller supply the PDU.

cvinayak committed:
Bluetooth: Controller: Refactor Periodic Advertising time update

Refactor Periodic Advertising time update function to use
caller supplied Periodic Advertising PDU to calculate the
time reservations.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@ppryga  I have included the refactoring in this PR now, in commit https://github.com/zephyrproject-rtos/zephyr/pull/36444/commits/62e8871af10c5bd54d50f691d967c5d8383ecc0b

ppryga commented:
@cvinayak thank you!

PR Approved.
